File: gin/internal/handler/players/crafting_recipes.go
package handler_players

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"

	model "cards_project/internal/model/players"
	"cards_project/internal/handler"
)

type CraftingRecipeHandler struct { db *gorm.DB }

func NewCraftingRecipeHandler(db *gorm.DB) *CraftingRecipeHandler {
	return &CraftingRecipeHandler{db: db}
}

func (h *CraftingRecipeHandler) RegisterRoutes(r gin.IRouter) {
	g := r.Group("/api/crafting_recipes")
	g.GET("", h.List)
	g.POST("", h.Create)
	g.GET("/:id", h.Get)
	g.PUT("/:id", h.Update)
	g.PATCH("/:id", h.Patch)
	g.DELETE("/:id", h.Delete)
	g.GET("/:id/api/crafting-recipes/{id}/can-craft", h.CanCraft)
	g.POST("/:id/api/crafting-recipes/{id}/craft", h.ExecuteCraft)
	g.POST("/:id/api/crafting-recipes/{id}/disable", h.Disable)
	g.POST("/:id/api/crafting-recipes/{id}/enable", h.Enable)
}

func (h *CraftingRecipeHandler) List(c *gin.Context) {
	skip, limit := handler.Paginate(c)
	var rows []model.CraftingRecipe
	if err := h.db.Offset(skip).Limit(limit).Find(&rows).Error; err != nil {
		handler.DbError(c, err); return
	}
	out := make([]model.CraftingRecipeResponse, len(rows))
	for i := range rows { out[i] = rows[i].ToResponse() }
	c.JSON(http.StatusOK, out)
}

func (h *CraftingRecipeHandler) Create(c *gin.Context) {
	var req model.CraftingRecipeCreateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		handler.ValidationError(c, err.Error()); return
	}
	if msgs := validateCraftingRecipe(&req); len(msgs) > 0 {
		c.JSON(http.StatusBadRequest, gin.H{"errors": msgs}); return
	}
	row := model.CraftingRecipe{}
	row.DustCost = req.DustCost
	row.IsAvailable = req.IsAvailable
	row.ResultCardID = req.ResultCardID
	if err := h.db.Create(&row).Error; err != nil {
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusCreated, row.ToResponse())
}

func (h *CraftingRecipeHandler) Get(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.CraftingRecipe
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "CraftingRecipe"); return }
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusOK, row.ToResponse())
}

func (h *CraftingRecipeHandler) Update(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.CraftingRecipe
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "CraftingRecipe"); return }
		handler.DbError(c, err); return
	}
	var req model.CraftingRecipeUpdateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		handler.ValidationError(c, err.Error()); return
	}
	row.ApplyUpdate(req)
	createReq := toCreateRequestCraftingRecipe(&row)
	if msgs := validateCraftingRecipe(&createReq); len(msgs) > 0 {
		c.JSON(http.StatusBadRequest, gin.H{"errors": msgs}); return
	}
	if err := h.db.Save(&row).Error; err != nil {
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusOK, row.ToResponse())
}

func (h *CraftingRecipeHandler) Patch(c *gin.Context) { h.Update(c) }

func (h *CraftingRecipeHandler) Delete(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	if err := h.db.Delete(&model.CraftingRecipe{}, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "CraftingRecipe"); return }
		handler.DbError(c, err); return
	}
	c.Status(http.StatusNoContent)
}

func (h *CraftingRecipeHandler) CanCraft(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.CraftingRecipe
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "CraftingRecipe"); return }
		handler.DbError(c, err); return
	}
	var body map[string]interface{}
	_ = c.ShouldBindJSON(&body)
	playerId := func() int {
		v, ok := body["player_id"]; if !ok { return 0 }
		f, ok := v.(float64); if !ok { return 0 }
		return int(f)
	}()
	result, err := row.CanCraft(playerId)
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.JSON(http.StatusOK, gin.H{"result": result})
}

func (h *CraftingRecipeHandler) ExecuteCraft(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.CraftingRecipe
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "CraftingRecipe"); return }
		handler.DbError(c, err); return
	}
	var body map[string]interface{}
	_ = c.ShouldBindJSON(&body)
	playerId := func() int {
		v, ok := body["player_id"]; if !ok { return 0 }
		f, ok := v.(float64); if !ok { return 0 }
		return int(f)
	}()
	err := row.ExecuteCraft(playerId)
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.Status(http.StatusNoContent)
}

func (h *CraftingRecipeHandler) Disable(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.CraftingRecipe
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "CraftingRecipe"); return }
		handler.DbError(c, err); return
	}
	err := row.Disable()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.Status(http.StatusNoContent)
}

func (h *CraftingRecipeHandler) Enable(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.CraftingRecipe
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "CraftingRecipe"); return }
		handler.DbError(c, err); return
	}
	err := row.Enable()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.Status(http.StatusNoContent)
}

func validateCraftingRecipe(req *model.CraftingRecipeCreateRequest) []string {
	var errs []string
	if !(req.DustCost > 0) {
		errs = append(errs, "Crafting recipe must have a dust cost greater than zero")
	}
	return errs
}

func toCreateRequestCraftingRecipe(m *model.CraftingRecipe) model.CraftingRecipeCreateRequest {
	return model.CraftingRecipeCreateRequest{
		DustCost: m.DustCost,
		IsAvailable: m.IsAvailable,
		ResultCardID: m.ResultCardID,
	}
}

File: express/src/routes/Players/crafting_recipes.ts
import { Router } from 'express';
import { prisma } from '../../lib/prisma.js';
import { CraftingRecipeService } from '../../services/Players/crafting_recipe_service.js';

const router = Router();
const service = new CraftingRecipeService();

function validate(data: any): void {
  if (!((data.dustCost == null || data.dustCost > 0))) throw new Error(`Crafting recipe must have a dust cost greater than zero`);
}

router.get('/', async (_req, res) => {
  const items = await prisma.craftingRecipe.findMany();
  res.json(items);
});

router.post('/', async (req, res) => {
  const body = req.body;
  const data: any = {};
    if (body.dustCost !== undefined) data.dustCost = body.dustCost;
    if (body.isAvailable !== undefined) data.isAvailable = body.isAvailable;
    if (body.resultCardId !== undefined) data.resultCardId = body.resultCardId;
  try {
  validate(data);
    const entity = await prisma.craftingRecipe.create({ data });
    res.status(201).json(entity);
  } catch (err: any) {
    res.status(400).json({ error: err?.message ?? 'Validation error' });
  }
});

router.get('/:id', async (req, res) => {
  const entity = await prisma.craftingRecipe.findUnique({ where: { id: Number(req.params.id) } });
  if (!entity) return res.status(404).json({ error: 'Not found' });
  res.json(entity);
});

router.put('/:id', async (req, res) => {
  const body = req.body;
  const data: any = {};
    if (body.dustCost !== undefined) data.dustCost = body.dustCost;
    if (body.isAvailable !== undefined) data.isAvailable = body.isAvailable;
    if (body.resultCardId !== undefined) data.resultCardId = body.resultCardId;
  try {
  validate(data);
    const entity = await prisma.craftingRecipe.update({ where: { id: Number(req.params.id) }, data });
    res.json(entity);
  } catch (err: any) {
    const status = err?.code === 'P2025' ? 404 : 400;
    res.status(status).json({ error: err?.message ?? 'Error' });
  }
});

router.patch('/:id', async (req, res) => {
  const body = req.body;
  const data: any = {};
    if (body.dustCost !== undefined) data.dustCost = body.dustCost;
    if (body.isAvailable !== undefined) data.isAvailable = body.isAvailable;
    if (body.resultCardId !== undefined) data.resultCardId = body.resultCardId;
  try {
  validate(data);
    const entity = await prisma.craftingRecipe.update({ where: { id: Number(req.params.id) }, data });
    res.json(entity);
  } catch (err: any) {
    const status = err?.code === 'P2025' ? 404 : 400;
    res.status(status).json({ error: err?.message ?? 'Error' });
  }
});

router.delete('/:id', async (req, res) => {
  try {
    await prisma.craftingRecipe.delete({ where: { id: Number(req.params.id) } });
    res.status(204).send();
  } catch {
    res.status(404).json({ error: 'Not found' });
  }
});

router.get('/:id/can-craft', async (req, res) => {
  const id = Number((req.params as any).id);
  const playerId = (req.query as any).playerId;
  try {
    const result = await service.can_craft(id, playerId);
    res.json({ result });
  } catch (err: any) {
    res.status(404).json({ error: err?.message ?? 'Not found' });
  }
});

router.post('/:id/craft', async (req, res) => {
  const id = Number((req.params as any).id);
  const playerId = req.body.playerId;
  try {
    await service.execute_craft(id, playerId);
    res.status(204).send();
  } catch (err: any) {
    res.status(404).json({ error: err?.message ?? 'Not found' });
  }
});

router.post('/:id/disable', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    await service.disable(id);
    res.status(204).send();
  } catch (err: any) {
    res.status(404).json({ error: err?.message ?? 'Not found' });
  }
});

router.post('/:id/enable', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    await service.enable(id);
    res.status(204).send();
  } catch (err: any) {
    res.status(404).json({ error: err?.message ?? 'Not found' });
  }
});
export default router;
